### PR TITLE
Add OAuth2 Authentication provider

### PR DESCRIPTION
**Pull request by nielsm5**

### CODE REVIEW

File: console/war/src/main/java/nl/nn/adapterframework/web/SecurityChainConfigurer.java [hunk 1]
```diff
@@ -78,6 +78,7 @@ public SecurityFilterChain configureChain(HttpSecurity http) {
 		ServletRegistration servlet = applicationContext.getBean("backendServletBean", ServletRegistration.class);
 		IAuthenticator authenticator = createAuthenticator();
 		authenticator.registerServlet(servlet.getServletConfiguration());
+		//TODO: register frontend here as well
```
Comment (nielsm5) [inline console/war/src/main/java/nl/nn/adapterframework/web/SecurityChainConfigurer.java:81]: #5416

File: security/src/main/java/nl/nn/adapterframework/lifecycle/servlets/ActiveDirectoryAuthenticator.java [hunk 2]
```diff
@@ -98,50 +88,11 @@ public SecurityFilterChain configure(HttpSecurity http) throws Exception {
...
-		provider.setAuthoritiesMapper(new LdapAuthorityMapper(roleMappingURL));
+		provider.setAuthoritiesMapper(new AuthorityMapper(roleMappingURL, getSecurityRoles(), getEnvironmentProperties()));
 
 		http.authenticationProvider(provider);
 		String realmName = StringUtils.isNotEmpty(domainName) ? domainName : url;
 		http.httpBasic().realmName(realmName);
 		return http.build();
 	}
-
-	public class LdapAuthorityMapper implements GrantedAuthoritiesMapper {
```
Comment (nielsm5) [inline security/src/main/java/nl/nn/adapterframework/lifecycle/servlets/ActiveDirectoryAuthenticator.java:109]: Has been moved to it's own custom class.

File: security/src/main/java/nl/nn/adapterframework/lifecycle/servlets/OAuth2Authenticator.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+		builder.authorizationGrantType(AuthorizationGrantType.AUTHORIZATION_CODE);
+
+		builder.scope(StringUtil.split(scopes));
+		builder.authorizationUri(authorizationUri);
+		builder.tokenUri(tokenUri);
+		builder.jwkSetUri(jwkSetUri);
+		builder.issuerUri(issuerUri);
+		builder.userInfoUri(userInfoUri);
+		builder.userNameAttributeName(userNameAttributeName);
+		builder.clientName(name);
```
Comment (jkosternl) [inline security/src/main/java/nl/nn/adapterframework/lifecycle/servlets/OAuth2Authenticator.java:176]: This is a builder, which you can chain together, without using semicolons.

File: security/src/main/java/nl/nn/adapterframework/lifecycle/servlets/ActiveDirectoryAuthenticator.java [hunk 4]
```diff
@@ -16,34 +16,23 @@
 package nl.nn.adapterframework.lifecycle.servlets;
```
Comment (tnleeuw) [inline security/src/main/java/nl/nn/adapterframework/lifecycle/servlets/ActiveDirectoryAuthenticator.java:16]: Update the copyrights?

File: security/src/main/java/nl/nn/adapterframework/lifecycle/servlets/OAuth2Authenticator.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+import nl.nn.adapterframework.util.EnumUtils;
+import nl.nn.adapterframework.util.SpringUtils;
+import nl.nn.adapterframework.util.StringUtil;
+
+/**
+ * OAuth2 Authentication provider which contains 4 defaults (Google, GitHub,
+ * Facebook and Okta), as well as a custom setting which allows users to
+ * use their own IDP.
+ * 
+ * <p>
```
Comment (tnleeuw) [inline security/src/main/java/nl/nn/adapterframework/lifecycle/servlets/OAuth2Authenticator.java:50]: I cannot find where the role-mapping is documented, or the name of the role-mapping file.

Would this be the right JavaDoc for it, or elsewhere?

And where are the types of possible authenticators documented?

File: security/src/main/java/nl/nn/adapterframework/lifecycle/servlets/ActiveDirectoryAuthenticator.java (hunk 4 above)
Comment (nielsm5) [reply security/src/main/java/nl/nn/adapterframework/lifecycle/servlets/ActiveDirectoryAuthenticator.java:16]: Tja...

**Review by tnleeuw — APPROVED** [review]

**Pull request merged**